File: gradle-plugin/src/main/java/org/komapper/gradle/codegen/GenerateTask.java
package org.komapper.gradle.codegen;

import java.io.IOException;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.List;
import java.util.Objects;
import javax.inject.Inject;
import org.gradle.api.DefaultTask;
import org.gradle.api.tasks.TaskAction;
import org.komapper.codegen.CodeGenerator;
import org.komapper.codegen.MetadataReader;
import org.komapper.codegen.Table;

public class GenerateTask extends DefaultTask {

  private final Generator settings;

  @Inject
  public GenerateTask(Generator settings) {
    this.settings = Objects.requireNonNull(settings);
  }

  @TaskAction
  public void run() throws ClassNotFoundException, SQLException, IOException {
    List<Table> tables = read();
    generate(tables);
  }

  private List<Table> read() throws ClassNotFoundException, SQLException {
    var jdbc = settings.getJdbc();
    try (var connection = createConnection(jdbc)) {
      MetadataReader reader =
          new MetadataReader(
              settings.getEnquote().get(),
              connection.getMetaData(),
              settings.getCatalog().getOrNull(),
              settings.getSchemaPattern().getOrNull(),
              settings.getTableNamePattern().getOrNull(),
              settings.getTableTypes().get());
      return reader.read();
    }
  }

  private Connection createConnection(Jdbc jdbc) throws ClassNotFoundException, SQLException {
    String driverClassName = jdbc.getDriver().get();
    String url = jdbc.getUrl().get();
    String user = jdbc.getUser().get();
    String password = jdbc.getPassword().get();
    Class.forName(driverClassName);
    return DriverManager.getConnection(url, user, password);
  }

  private void generate(List<Table> tables) throws IOException {
    CodeGenerator generator =
        new CodeGenerator(
            settings.getPackageName().getOrNull(),
            tables,
            settings.getClassNameResolver().get(),
            settings.getPropertyNameResolver().get());
    var destinationDir = settings.getDestinationDir().get().getAsFile().toPath();
    try (var writer =
        generator.createNewFile(
            destinationDir, "entities.kt", settings.getOverwriteEntities().get())) {
      generator.generateEntities(
          writer,
          settings.getDeclareAsNullable().get(),
          settings.getUseSelfMapping().get(),
          settings.getUseCatalog().get(),
          settings.getUseSchema().get(),
          settings.getPropertyTypeResolver().get(),
          settings.getVersionPropertyName().get(),
          settings.getCreatedAtPropertyName().get(),
          settings.getUpdatedAtPropertyName().get());
    }
    if (!settings.getUseSelfMapping().get()) {
      try (var writer =
          generator.createNewFile(
              destinationDir, "entityDefinitions.kt", settings.getOverwriteDefinitions().get())) {
        generator.generateDefinitions(
            writer,
            settings.getUseCatalog().get(),
            settings.getUseSchema().get(),
            settings.getVersionPropertyName().get(),
            settings.getCreatedAtPropertyName().get(),
            settings.getUpdatedAtPropertyName().get());
      }
    }
  }
}

File: gradle-plugin/src/main/kotlin/org/komapper/gradle/codegen/GenerateTask.kt
package org.komapper.gradle.codegen

import org.gradle.api.DefaultTask
import org.gradle.api.tasks.TaskAction
import org.komapper.codegen.CodeGenerator
import org.komapper.jdbc.JdbcDatabase
import org.komapper.jdbc.dsl.MetadataDsl.tables
import org.komapper.jdbc.dsl.query.MetadataQuery
import javax.inject.Inject

open class GenerateTask @Inject internal constructor(private val settings: Generator) : DefaultTask() {

    @TaskAction
    fun run() {
        val database = settings.database.get()
        val tables = read(database)
        generate(tables)
    }

    private fun read(database: JdbcDatabase): List<MetadataQuery.Table> {
        val metadataQuery = tables(
            catalog = settings.catalog.orNull,
            schemaPattern = settings.schemaPattern.orNull,
            tableNamePattern = settings.tableNamePattern.orNull,
            tableTypes = settings.tableTypes.get()
        )
        return metadataQuery.run(database.config)
    }

    private fun generate(tables: List<MetadataQuery.Table>) {
        val generator = CodeGenerator(
            destinationDir = settings.destinationDir.get().asFile.toPath(),
            packageName = settings.packageName.orNull,
            prefix = settings.prefix.get(),
            suffix = settings.suffix.get(),
            tables = tables
        )
        generator.generateEntities(
            resolver = settings.classResolver.get(),
            overwrite = settings.overwriteEntities.get(),
            declareAsNullable = settings.declareAsNullable.get()
        )
        generator.generateDefinitions(
            overwrite = settings.overwriteDefinitions.get(),
            useCatalog = settings.useCatalog.get(),
            useSchema = settings.useSchema.get()
        )
    }
}
